cli/flox-rust-sdk/src/providers/build.rs

```diff
@@ -1010,6 +1010,127 @@ mod tests {
         );
     }
 
+    #[test]
+    fn build_impure_patches_binaries() {
+        let package_name = String::from("foo");
+        let file_name = String::from("links-against-libc");
+
+        // Force CGO without providing `gcc` from the Flox environment.
+        //
+        // In reality `gcc` and `libc` will come from the `devShell` rather than
+        // a host OS which complicates these tests and means that we can never
+        // trust `rpath` or `runpath`.
```

Comment:
Not sure if I understand the trust issue here.
Is this influenced by environment vars and can we perhaps make this more reliable by using `temp_env` then?

Reply:
"trust" might have been too stronger word and it's more of an issue for the interpreter than rpath/runpath: https://github.com/flox/flox/pull/2234#discussion_r1797111308

Any time that `gcc` and `libc` come Nix (via devShell or profile, rather than some other Linux system package) then the interpreter and rpath/runpath will point at store paths. That makes it impossible to observe the effect of the `autoPatchelfHook` by checking whether they point at store paths.

I tried unsetting `NIX_ENFORCE_NO_NATIVE`, `NIX_LDFLAGS`, and `NIX_CC` to no effect. I suspect that we'd need to revert `PATH` to prior to the `devShell` but then that would cause the tests to fail on our builders which don't have a system-wide `gcc` and `glibc`.

Reply:
so,.. the issue is that go uses the c compiler it finds on the PATH which may be a system provided one, but while nix wraps the compiler with additional nix flags, when running the test within a nix shell, we don't see this fail?

I wonder if putting gcc-unwrapped on top of the PATH would help, but i figure that just fails to compile without also pointing to the right stdlibs

Reply:
> I suspect that we'd need to revert PATH to prior to the devShell but then that would cause the tests to fail on our builders which don't have a system-wide gcc and glibc

One way to ensure a consistent experience would be to reset the PATH to contain _only_ the build environment as we invoke the build script. That would of course impose the requirement that gcc be installed to that build environment, but is that not what we're looking for?

(Relatedly, I've been chatting with @stahnma about adding a `flox activate --pure` flag which would set the PATH to be only the flox-provided bin and sbin directories. If we had this flag then we could add it to the build script invocation on line 202 of flox-build.mk.)

Reply:
yeah doing impure builds and expecting a pure result dont mix too well...

> One way to ensure a consistent experience would be to reset the PATH to contain only the build environment as we invoke the build script. That would of course impose the requirement that gcc be installed to that build environment, but is that not what we're looking for?

It's certainly what I would be looking for but impure builds seem to explicitly allow for that.
A `pure` flag would be nice but i suspect that would rub badly with the semantics of `sandbox = "pure"`.

Reply:
Should we skip the test if `gcc` is coming from `/nix/store`?